test/clj/rems/api/services/test_invitation.clj

```diff
@@ -0,0 +1,113 @@
+(ns ^:integration rems.api.services.test-invitation
+  (:require [clojure.test :refer :all]
+            [rems.api.services.invitation :as invitation]
+            [rems.db.test-data-helpers :as test-helpers]
+            [rems.db.testing :refer [caches-fixture rollback-db-fixture test-db-fixture]]
+            [rems.email.core :as email]
+            [rems.testing-util :refer [fixed-time-fixture with-user]])
+  (:import [org.joda.time DateTime DateTimeUtils DateTimeZone]))
+
+(def invitation-time (DateTime. 10000 DateTimeZone/UTC))
+(def acceptance-time (DateTime. 20000 DateTimeZone/UTC))
+
+(use-fixtures
+  :once
+  test-db-fixture
+  caches-fixture
+  (fixed-time-fixture invitation-time))
+
+(use-fixtures :each rollback-db-fixture)
+
+(deftest test-crud-invitation
+  (test-helpers/create-user! {:eppn "owner" :commonName "owner" :mail "owner@example.com"} :owner)
+  (test-helpers/create-user! {:eppn "new-handler" :commonName "New Handler" :mail "new-handler@example.com"})
+  (let [workflow-id (test-helpers/create-workflow! {})]
+    (with-user "owner"
+      (testing "before creating invitations"
+        (is (= [] (invitation/get-invitations nil))))
+
+      (testing "creating invitations"
+        (testing "without any type"
+          (is (= {:success false
+                  :errors
+                  [{:type :t.accept-invitation.errors/invalid-invitation-type :workflow-id nil}]}
+                 (invitation/create-invitation! {:userid "owner"
+                                                 :name "Dorothy Vaughan"
+                                                 :email "dorothy.vaughan@nasa.gov"}))))
```

Comment:
i guess this test refers to performing invitation without `workflow-id`? if so, perhaps this could be better reflected in the test name

Reply:
My thinking is this:
- The actual test name refers to the CRUD functionality which is a collection of create, read/list, then modify, then ... that are interdependent so they work best in a single larger test function.
- Inside, we try creating an invitation without a type, which is an error.
- The only currently supported type of invitation (in this "v2" invitation PR) is a workflow handler invitation.
- There exists a "v1" invitation feature for application members and reviewers that may or may not be ported to this new one.
- In the future, with several invitation types (e.g. owner) there would also be `:userid nil` in the error. It shows you all the values it checked for the "type".
- I chose to have the test try to have an otherwise valid invitation (to a workflow) so that the only thing wrong is the missing type.

Reply:
check, thanks for the explanation! i was trying to find a `:type` in the map and maybe that confused me a little bit. :)